engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala

```diff
@@ -0,0 +1,253 @@
+  /*
+    * Create 2 Tables, one for keys in done status and one for keys in terminal status.
+    * A Table is nothing more than a Map[R, Map[C, V]], see Table trait for more details
+    * In this case, rows are GraphNodes, columns are ExecutionIndexes, and values are JobKeys
+    * This allows for quick lookup of all shards for a node, as well as accessing a specific key with a 
+    * (node, index) pair
+   */
+  lazy val (doneStatus, terminalStatus) = {
+    def toTableEntry(key: JobKey) = (key.node, key.index, key)
+
+    store.foldLeft((Table.empty[GraphNode, ExecutionIndex, JobKey], Table.empty[GraphNode, ExecutionIndex, JobKey]))({
+      case ((done, terminal), (status, keys))  =>
+        lazy val newMapEntries = keys map toTableEntry
+        val newDone = if (status.isDoneOrBypassed) done.addAll(newMapEntries) else done
+        val newTerminal = if (status.isTerminal) terminal.addAll(newMapEntries) else terminal
+
+        newDone -> newTerminal
+    })
+  }
+
+  private def keysWithStatus(status: ExecutionStatus) = store.getOrElse(status, List.empty)
+
+  def isInBypassedConditional(jobKey: JobKey): Boolean = keysWithStatus(Bypassed).exists {
+    case conditional: ConditionalKey if conditional.node.innerGraph.nodes.contains(jobKey.node)
+      && conditional.index.equals(jobKey.index) => true
+    case _ => false
+  }
+
+  /**
+    * We're done when all the keys have a terminal status,
+    * which is equivalent to non of them being in a non-terminal status and faster to verify
+    */
+  def isDone: Boolean = {
+    NonTerminalStatuses.map(keysWithStatus).forall(_.isEmpty)
+  }
+
+  def jobStatus(jobKey: JobKey): Option[ExecutionStatus] = statusStore.get(jobKey)
+
+  def startedJobs: List[BackendJobDescriptorKey] = {
+    store.filterNot({ case (s, _) => s == NotStarted}).values.toList.flatten collect {
+      case k: BackendJobDescriptorKey => k
+    }
+  }
+
+  override def toString: String = store.map { case (j, s) => s"$j -> $s" } mkString System.lineSeparator()
+
+  /**
+    * If needsUpdate is true, goes through NotStarted keys and determines the ones that can be run.
+    * Only computes the first MaxJobsToStartPerTick runnable keys.
+    * In the process, identifies and updates the status of keys that are unreachable.
+    * Returns an ExecutionStoreUpdate which is the list of runnable keys and an updated execution store.
+    * 
+    * If needsUpdate, returns an empty list of runnable keys and this instance of the store.
+    * 
+    * This method can expansive to run for very large workflows if needsUpdate is true.
+    */
+  def update: ExecutionStoreUpdate = if (needsUpdate) {
+    // When looking for runnable keys, keep track of the ones that are unreachable so we can mark them as such
+    var unreachables = Map.empty[JobKey, ExecutionStatus]
+
+    // filter the keys that are runnable. In the process remember the ones that are unreachable
+    val readyToStart = keysWithStatus(NotStarted).toStream.filter(key => {
```

Comment:
ToL curious why `toStream`?

Reply:
Two things:
- To avoid making Cromwell cry too much, a little while ago we limited the number of "runnable keys" that this function will return per call so that if you have a 1M wide scatter, Cromwell doesn't starts 1M job literally at the same time. Instead, we return up to the first 1000 ones. This is called every second so it's still relatively fast but give some breathing room to Cromwell.

- It's potentially expensive to determine if all the dependencies of a key are in a given status (it's actually less now than it used to be since we've optimized the data structures but...).

Because of those 2 things, I wanted to avoid calculating "are you runnable?" for all of the keys if we're only going to take the first 1000 anyway. A stream / filter / take lazily computes only the first 1000 runnable keys and then stops, so we don't lose any cycle on the rest.

If you wonder why all this fuss, we had performance issues at a few different occasions as people started throwing bigger and bigger scatters to Cromwell and tracked it down to lots of CPU spent computing runnable keys.

Reply:
Very interesting, thanks for the in-depth explanation!  

In the interest of future me forgetting all this, could you please paste this explanation into a comment?